File: 37.sudoku-solver.py
class Solution:
    def solve(self, board, state, i_begin, j_begin):
        for i in range(i_begin, 9):
            for j in range(i_begin, 9):
                if board[i][j] == '.':
                    pos = j // 3 * 3 + i // 3
                    for n in range(9):
                        if state[0][i][n] == False and state[1][j][n] == False and state[2][pos][n] == False:
                            board[i][j] = str(1 + n)
                            state[0][i][n] = True
                            state[1][j][n] = True
                            state[2][pos][n] = True
                            if self.solve(board, state, i_begin, j_begin):
                                return True
                            state[0][i][n] = False
                            state[1][j][n] = False
                            state[2][pos][n] = False
                    board[i][j] = '.'
                    return False
        return True;

    def solveSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: void Do not return anything, modify board in-place instead.
        """
        state = [[[False for _ in range(9)] for _ in range(9)] for _ in range(3)]
        for i in range(9):
            for j in range(9):
                if board[i][j] != '.':
                    pos = ord(board[i][j]) - ord('1')
                    state[0][i][pos] = True
                    state[1][j][pos] = True
                    state[2][j // 3 * 3 + i // 3][pos] = True
        self.solve(board, state, 0, 0)

File: 37.sudoku-solver.cpp
class Solution {
public:
    bool solve(vector<vector<char>>& board, array<array<array<bool, 9>, 9>, 3>& state, int i_begin, int j_begin) {
        for (int i = i_begin; i < 9; i++) {
            for (int j = j_begin; j < 9; j++) {
                if (board[i][j] == '.') {
                    for (int n = 0; n < 9; n++) {
                        if(state[0][i][n] == false && state[1][j][n] == false && state[2][j / 3 * 3 + i / 3][n] == false) {
                            board[i][j] = '1' + n;
                            state[0][i][n] = true;
                            state[1][j][n] = true;
                            state[2][j / 3 * 3 + i / 3][n] = true;
                            if (solve(board, state, i_begin, j_begin)) {
                                return true;
                            }
                            state[0][i][n] = false;
                            state[1][j][n] = false;
                            state[2][j / 3 * 3 + i / 3][n] = false;
                        }
                    }
                    board[i][j] = '.';
                    return false;
                }
            }
        }
        return true;
    }

    void solveSudoku(vector<vector<char>>& board) {
        array<array<array<bool, 9>, 9>, 3> state;
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                state[0][i][j] = false;
                state[1][i][j] = false;
                state[2][i][j] = false;
            }
        }
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] != '.') {
                    int pos = board[i][j] - '1';
                    state[0][i][pos] = true;
                    state[1][j][pos] = true;
                    state[2][j / 3 * 3 + i / 3][pos] = true;
                }
            }
        }
        solve(board, state, 0, 0);
    }
};